### PR TITLE
feat(model_runner): remove pool/backend refs from ForwardBatch via ForwardContext

### PR DESCRIPTION
**Pull request by ch-wan**
## Motivation

`ForwardBatch` carried four runtime-ref fields — `req_to_token_pool`, `token_to_kv_pool`, `attn_backend`, `hisparse_coordinator` — that have nothing to do with the per-batch input data the dataclass exists for. They were threaded into every model-layer file (`forward_mha.py`, `forward_mla.py`, `qwen3.py`, `deepseek_v4.py`, …), every attention backend, every hand-rolled `ForwardBatch(...)` constructor, and every speculative-decoding path. Removing them lets `ForwardBatch` stay focused on actual batch state, makes the attention backend solely responsible for its own pool refs, and gives PDmux / TBO / spec a uniform way to swap the active backend.

## Modifications

- New `python/sglang/srt/model_executor/forward_context.py` exporting a frozen `ForwardContext` dataclass (single mandatory field `attn_backend`) plus a scoped `forward_context()` context manager. Convenience pool accessors `get_token_to_kv_pool()` / `get_req_to_token_pool()` derive from `get_attn_backend().*` — no separate pool global needed.
- Drop four fields from `ForwardBatch`: `req_to_token_pool`, `token_to_kv_pool`, `attn_backend`, `hisparse_coordinator`. Drop the matching assignments from `ForwardBatch.init_new()` and from every hand-rolled `ForwardBatch(...)` constructor in graph runners, speculative graph runners, `cpu_graph_runner`, `dflash_worker`, and `two_batch_overlap`.
- **Pattern A** (~21 attention backends + `dsa_indexer` / `dsv4_compressor` family): cache `self.req_to_token_pool` / `self.token_to_kv_pool` at `__init__(model_runner)`. Reads inside backend bodies switch from `forward_batch.<pool>` to `self.<pool>`.
- **`attn_backend` dispatch**: `ModelRunner._forward_raw` publishes `ForwardContext(attn_backend=self.attn_backend)` at entry with save/restore via `forward_context()`. PDmux per-stream, TBO per-child, and spec-draft per-step overrides use `dataclasses.replace(get_forward_context(), attn_backend=…)` to nest a child context. `RadixAttention.forward` / `RadixLinearAttention.forward` dispatch via `get_attn_backend().forward(...)`.
- **`hisparse_coordinator`**: backend `__init__` captures from `model_runner` where available; for the DSV4/DSA family the coordinator is constructed after `init_attention_backend()`, so those backends expose it as a `@property` reading `self.model_runner.hisparse_coordinator` live (consumed by `C4IndexerBackendMixin`).
- **`start_layer`** (Pattern B, already on `main` via PR #25825): no further changes — model-layer code reads `self.start_layer` injected at decoder-layer construction.
- **Frozen-KV MTP**: `target_kv_pool_view` simplifies to a single swap of `draft_attn_backend.token_to_kv_pool` for the scope. Both readers see it — `get_token_to_kv_pool()` resolves through `get_attn_backend()`, and the backend's own `self.token_to_kv_pool` is the same object. The previous three-way swap (runner attr + backend attr + pool global) collapses to one.
- **`ForwardBatchDeepSeekMHAMixin`** (`prepare_chunked_prefix_cache_info` etc.): switched to `get_req_to_token_pool()` / `get_token_to_kv_pool()` (was still reading the deleted fields off `ForwardBatch`).
- **Latent bug fixed**: `BaseSWAKVPool` and `DeepSeekV4TokenToKVPool` were missing `invalidate_loc_cache()`; `SWATokenToKVPoolAllocator.clear()` calls it unconditionally and would have raised `AttributeError` on the DSV4 path. Default no-op added to `BaseSWAKVPool`; real override on `DeepSeekV4TokenToKVPool` clears its `cached_loc`.

After the change: `grep -rn 'forward_batch\.\(token_to_kv_pool\|req_to_token_pool\|attn_backend\|hisparse_coordinator\)' python/sglang/srt/ test/ --include='*.py' | grep -v forward_batch_info.py` returns zero matches.

## Accuracy Tests

The migrated callsites are mechanical — every `forward_batch.<pool>` resolves to the same object that `get_attn_backend().<pool>` / `self.<pool>` now returns. The risk surface is multi-runner (target + spec draft) where the wrong pool would silently land target writes in the draft buffer; the `ForwardContext` save/restore in `_forward_raw` defends against that. Cluster validation matrix is in flight (DSV4-Flash FP4 B200 GSM8K-gated EAGLE test + variants); GSM8K accuracy gate at 0.93 catches silent pool corruption.

## Speed Tests and Profiling

No hot-path overhead. Each `_forward_raw` adds one frozen-dataclass construction and one global pointer swap (with restore) per forward; readers replace an `obj.attr` lookup with `obj.attr.attr` (one extra pointer hop) for callers that previously hit `forward_batch.<pool>`. No new allocations on the steady-state decode path.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #26252835332](https://github.com/sgl-project/sglang/actions/runs/26252835332)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #26252834888](https://github.com/sgl-project/sglang/actions/runs/26252834888)

### CODE REVIEW

File: python/sglang/srt/model_executor/breakable_cuda_graph_runner.py [hunk 1]
```diff
@@ -459,6 +463,9 @@ def replay_layer_forward(*args, **layer_kwargs):
             original_layer_forward = self.layer_model.forward
             self.layer_model.forward = replay_layer_forward
             try:
+                # No forward_context wrap here: replay is called from
+                # ModelRunner._forward_raw which has already published one,
+                # and re-wrapping would clobber PDmux per-stream overrides.
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/breakable_cuda_graph_runner.py]: cleanup

File: python/sglang/srt/model_executor/forward_context.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+def get_forward_context() -> ForwardContext:
+    assert _current is not None, (
+        "no forward context active — call forward_context(...) or set_forward_context(...) "
+        "before reading get_forward_context()."
+    )
+    return _current
+
+
+def get_attn_backend() -> "AttentionBackend":
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/forward_context.py]: quotes are not needed

File: python/sglang/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -3000,14 +3008,25 @@ def forward_decode(
...
             )
 
+        with ctx:
+            if pdmux_override:
+                with forward_context(
+                    replace(
+                        get_forward_context(), attn_backend=self.decode_attn_backend
+                    )
+                ):
+                    return _do_forward()
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner.py:3024]: Can we simply create a new ForwardContext rather than using hacky `replace` here?

File: python/sglang/srt/speculative/frozen_kv_mtp_worker.py [hunk 4]
```diff
@@ -177,6 +178,9 @@ def __init__(
 
         self.draft_attn_backend = self._init_draft_attn_backend()
         self.draft_model_runner.draft_attn_backend = self.draft_attn_backend
+        # For topk > 1 the draft backend differs from the runner's main
```
Comment (ch-wan) [inline python/sglang/srt/speculative/frozen_kv_mtp_worker.py]: remove comments here

File: python/sglang/srt/speculative/frozen_kv_mtp_worker.py [hunk 5]
```diff
@@ -290,7 +299,8 @@ def _init_frozen_kv_metadata_capture_cuda_graph(
                 forward_mode=ForwardMode.DECODE,
                 spec_info=None,
             )
-        forward_batch.attn_backend = self.draft_attn_backend
+        # Each draft forward call wraps with forward_context to publish
```
Comment (ch-wan) [inline python/sglang/srt/speculative/frozen_kv_mtp_worker.py]: remove

File: python/sglang/srt/speculative/frozen_kv_mtp_worker.py [hunk 6]
```diff
@@ -275,7 +283,8 @@ def _init_frozen_kv_metadata(self, forward_batch: ForwardBatch) -> None:
             forward_batch.seq_lens_sum = torch.sum(forward_batch.seq_lens).item()
         with self._frozen_kv_target_view(forward_batch):
             self.draft_attn_backend.init_forward_metadata(forward_batch)
-        forward_batch.attn_backend = self.draft_attn_backend
+        # Each draft forward call wraps with forward_context to publish
```
Comment (ch-wan) [inline python/sglang/srt/speculative/frozen_kv_mtp_worker.py]: remove

File: python/sglang/srt/speculative/frozen_kv_mtp_worker.py [hunk 7]
```diff
@@ -310,7 +320,8 @@ def _init_frozen_kv_metadata_replay_cuda_graph(
                     else None
                 ),
             )
-        forward_batch.attn_backend = self.draft_attn_backend
+        # Each draft forward call wraps with forward_context to publish
+        # self.draft_attn_backend, so no FB or runner mutation is needed.
```
Comment (ch-wan) [inline python/sglang/srt/speculative/frozen_kv_mtp_worker.py]: remove
